SIPS/sip-12.md

```diff
@@ -0,0 +1,114 @@
+
+### Definitions
+
+> This section is non-normative, and merely recapitulates some definitions from [CAIP-2](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-2.md).
+
+- `ChainId` - a [CAIP-2](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-2.md) string.
+  It identifies a specific chain among all blockchains recognized by the CAIP standards.
+  - `ChainId` consists of a `Namespace` and a `Reference`
+    - `Namespace` - a class of similar blockchains. For example EVM-based blockchains.
+    - `Reference` - a way to identify a concrete chain inside a `Namespace`. For example Ethereum Mainnet or one of its test networks.
+
+- `AccountAddress` - The account address portion of a [CAIP-10](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-10.md) ID.
+
+### Snap Manifest
+
+This SIP specifies a permission named `endowment:name-lookup`.
+The permission grants a snap the ability to expose an `onNameLookup` export that receives an object with `chainId` and `domain` OR `address` fields.
+
+This permission is specified as follows in `snap.manifest.json` files:
+
+```json
+{
+  "initialPermissions": {
+    "endowment:name-lookup": {
+        "chains": ["eip155:1", "bip122:000000000019d6689c085ae165831e93"],
+    }
+  }
+}
+```
+
+`chains` - An array of CAIP-2 chain IDs that the snap supports. This field is useful for a client in order to avoid unnecessary overhead.
+
+### Snap Implementation
+
+Any snap that wishes to provide name lookup features **MUST** implement the following API:
+
+```typescript
+import { OnNameLookupHandler } from "@metamask/snap-types";
+
+export const onNameLookup: OnNameLookupHandler = async ({
+  chainId,
+  domain,
+  address
+}) => {
+  let resolution;
+  if (domain) {
+    const getAddress = (domain) => /* Get address */;
+    resolution = getAddress(domain);
+  } else if (address) {
+    const getDomain = (address) => /* Get domain */;
+    resolution = getDomain(address);
+  } else {
+    return null;
+  }
+
+  return { resolution };
+};
+```
+
+The type for an `onNameLookup` handler function’s arguments is:
+
+```typescript
+type OnNameLookupArgs = {
+  chainId: Caip2ChainId;
+} & ({ domain: string, address: undefined } | { address: string, domain: undefined });
+
+```
+
+`chainId` - This is a [CAIP-2](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-2.md) `chainId` string.
+The snap is expected to parse and utilize this string as needed.
+
+`domain` - This is a human-readable address. If the `domain` property is defined, the request is looking for resolution to an address.
+
+`address` - This is a non-readable address, this should be the native address format of the currently selected chain/protocol. If the `address` property is defined,
+the request is looking for resolution to a domain.
+
+The interface for the return value of an `onNameLookup` export is:
+
+```typescript
+type Domain = string;
+
+type OnNameLookupResponse = {
+  resolution: AccountAddress | Domain;
```

Comment:
This could allow for cleaner implementations:

```suggestion
  resolution: AccountAddress | Domain | null;
```

Then if you do something like this...

```ts
return { resolution: getAddress(...); }
```

...the `getAddress` function could return `null`. Now, you'd have to work around it like

```ts
const address = getAddress(...);
if (address) {
  return { resolution: address };
}

return null;
```